File: frontend/src/api/video.js
import api from './axios';

/**
 * 영상 관련 API 함수 모음.
 *
 * 서버 엔드포인트:
 *   GET    /api/playlists/{playlistId}/videos                       — 영상 목록 조회
 *   POST   /api/playlists/{playlistId}/videos                       — 영상 추가
 *   DELETE /api/playlists/{playlistId}/videos/{videoId}             — 영상 삭제
 *   PATCH  /api/playlists/{playlistId}/videos/{videoId}/volume      — 음량 업데이트
 */

/**
 * 플레이리스트의 영상 목록을 재생 순서(position ASC)로 조회한다.
 *
 * @param {number} playlistId 조회할 플레이리스트 ID
 * @returns {Promise<Array<{id, youtubeId, title, thumbnail, position, volume, createdAt}>>}
 */
export const getVideos = async (playlistId) => {
  const response = await api.get(`/playlists/${playlistId}/videos`);
  return response.data;
};

/**
 * 플레이리스트에 YouTube 영상을 추가한다.
 * 서버에서 YouTube API를 호출해 title/thumbnail을 자동으로 채운다.
 *
 * @param {number} playlistId  영상을 추가할 플레이리스트 ID
 * @param {string} youtubeUrl  유저가 입력한 YouTube URL
 * @returns {Promise<{id, youtubeId, title, thumbnail, position, volume, createdAt}>}
 * @throws 400 — 유효하지 않은 YouTube URL
 * @throws 404 — YouTube에 존재하지 않는 영상
 */
export const addVideo = async (playlistId, youtubeUrl) => {
  const response = await api.post(`/playlists/${playlistId}/videos`, { youtubeUrl });
  return response.data;
};

/**
 * 플레이리스트에서 영상을 삭제한다.
 *
 * @param {number} playlistId 플레이리스트 ID
 * @param {number} videoId    삭제할 영상 ID
 */
export const removeVideo = async (playlistId, videoId) => {
  await api.delete(`/playlists/${playlistId}/videos/${videoId}`);
};

/**
 * 영상의 음량을 업데이트한다.
 * 유저가 볼륨 슬라이더를 조절할 때 호출되어 DB에 즉시 반영된다.
 * 이후 해당 영상 재생 시 저장된 volume이 iframe에 자동 적용된다.
 *
 * @param {number} playlistId 플레이리스트 ID
 * @param {number} videoId    음량을 변경할 영상 ID
 * @param {number} volume     새 음량 (0~100)
 * @returns {Promise<{id, youtubeId, title, thumbnail, position, volume, createdAt}>}
 */
export const updateVolume = async (playlistId, videoId, volume) => {
  const response = await api.patch(
    `/playlists/${playlistId}/videos/${videoId}/volume`,
    { volume }
  );
  return response.data;
};

File: backend/src/main/java/com/youflow/video/Video.java
package com.youflow.video;

import com.youflow.playlist.Playlist;
import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * 영상(Video) 엔티티 — videos 테이블과 1:1 매핑.
 *
 * 관계 설계: Video → Playlist 단방향 @ManyToOne.
 * Playlist 쪽에 @OneToMany를 두지 않아 불필요한 컬렉션 로딩을 방지한다.
 *
 * 핵심 필드:
 *   - youtubeId: iframe 임베드에 사용하는 YouTube 영상 고유 ID
 *   - position:  플레이리스트 내 재생 순서 (0-based)
 *   - volume:    이 영상에 적용할 음량 (0~100), 기본값 70
 */
@Entity
@Table(name = "videos")
public class Video {

    /** DB auto-increment PK */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * 이 영상이 속한 플레이리스트.
     * LAZY 로딩: 영상 조회 시 Playlist 쿼리가 자동으로 나가지 않는다.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "playlist_id", nullable = false)
    private Playlist playlist;

    /** 유저가 입력한 원본 YouTube URL */
    @Column(name = "youtube_url", nullable = false)
    private String youtubeUrl;

    /**
     * YouTube 영상 고유 ID (URL에서 파싱).
     * iframe 임베드, YouTube Data API 호출 모두 이 값을 사용한다.
     * 예: https://www.youtube.com/watch?v=dQw4w9WgXcQ → dQw4w9WgXcQ
     */
    @Column(name = "youtube_id", nullable = false)
    private String youtubeId;

    /** YouTube Data API로 가져온 영상 제목 */
    @Column(nullable = false)
    private String title;

    /** YouTube Data API로 가져온 썸네일 URL */
    @Column(nullable = false)
    private String thumbnail;

    /**
     * 플레이리스트 내 재생 순서 (0-based).
     * 영상 추가 시 Service에서 현재 최대 position + 1로 설정한다.
     */
    @Column(nullable = false)
    private int position = 0;

    /**
     * 이 영상에 적용할 음량 (0~100).
     * 기본값 70. 유저가 조절하면 실시간으로 DB에 반영되고,
     * 다음 재생 시 iframe의 onReady에서 player.setVolume()으로 적용된다.
     */
    @Column(nullable = false)
    private int volume = 70;

    /** 영상 추가 시각 */
    @Column(nullable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    /** JPA 전용 기본 생성자 — 외부에서 직접 호출하지 않는다. */
    protected Video() {}

    /**
     * 영상 생성자. Service에서 YouTube API 응답을 파싱한 뒤 호출한다.
     *
     * @param playlist   영상이 추가될 플레이리스트
     * @param youtubeUrl 유저가 입력한 원본 URL
     * @param youtubeId  URL에서 파싱한 YouTube 영상 ID
     * @param title      YouTube API에서 가져온 영상 제목
     * @param thumbnail  YouTube API에서 가져온 썸네일 URL
     * @param position   플레이리스트 내 순서
     */
    public Video(Playlist playlist, String youtubeUrl, String youtubeId,
                 String title, String thumbnail, int position) {
        this.playlist = playlist;
        this.youtubeUrl = youtubeUrl;
        this.youtubeId = youtubeId;
        this.title = title;
        this.thumbnail = thumbnail;
        this.position = position;
    }

    public Long getId() { return id; }
    public Playlist getPlaylist() { return playlist; }
    public String getYoutubeUrl() { return youtubeUrl; }
    public String getYoutubeId() { return youtubeId; }
    public String getTitle() { return title; }
    public String getThumbnail() { return thumbnail; }
    public int getPosition() { return position; }
    public int getVolume() { return volume; }
    public LocalDateTime getCreatedAt() { return createdAt; }

    /** 음량 업데이트. WebSocket 또는 REST API로 유저가 조절할 때 호출한다. */
    public void updateVolume(int volume) {
        this.volume = volume;
    }

    /** 재생 순서 업데이트. 드래그-앤-드롭 재정렬 시 호출한다. */
    public void updatePosition(int position) {
        this.position = position;
    }
}
